File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/public/extensions/RunAnywhere+Storage.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public API for storage operations.
 * Provides storage information and management.
 *
 * Mirrors Swift RunAnywhere+Storage.swift pattern.
 */

package com.runanywhere.sdk.public.extensions

import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.Storage.StorageAvailability
import com.runanywhere.sdk.public.extensions.Storage.StorageInfo

// MARK: - Storage Information

/**
 * Get complete storage information.
 *
 * @return Storage info with device, app, and model storage details
 */
expect suspend fun RunAnywhere.storageInfo(): StorageInfo

/**
 * Check if storage is available for a download.
 *
 * @param requiredBytes Required bytes for the operation
 * @return Storage availability result
 */
expect suspend fun RunAnywhere.checkStorageAvailability(requiredBytes: Long): StorageAvailability

// MARK: - Cache Management

/**
 * Get cache size in bytes.
 *
 * @return Cache size
 */
expect suspend fun RunAnywhere.cacheSize(): Long

/**
 * Clear the SDK cache.
 */
expect suspend fun RunAnywhere.clearCache()

// MARK: - Storage Limits

/**
 * Set maximum storage limit for models.
 *
 * @param maxBytes Maximum bytes to use for model storage
 */
expect suspend fun RunAnywhere.setMaxModelStorage(maxBytes: Long)

/**
 * Get current storage used by models.
 *
 * @return Total bytes used by downloaded models
 */
expect suspend fun RunAnywhere.modelStorageUsed(): Long

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/Storage/RunAnywhere+Storage.swift
//
//  RunAnywhere+Storage.swift
//  RunAnywhere SDK
//
//  Public API for storage and download operations.
//

import Foundation

// MARK: - Model Download API

public extension RunAnywhere {

    /// Download a model by ID with progress tracking
    ///
    /// ```swift
    /// for await progress in try await RunAnywhere.downloadModel("my-model-id") {
    ///     print("Progress: \(Int(progress.overallProgress * 100))%")
    /// }
    /// ```
    static func downloadModel(_ modelId: String) async throws -> AsyncStream<DownloadProgress> {
        let logger = SDKLogger(category: "RunAnywhere.Download")
        let models = try await availableModels()
        logger.info("Available models count: \(models.count)")
        for m in models where m.id == modelId {
            logger.info("Found model \(m.id) with framework: \(m.framework.rawValue) (\(m.framework.displayName))")
        }
        guard let model = models.first(where: { $0.id == modelId }) else {
            throw SDKError.general(.modelNotFound, "Model not found: \(modelId)")
        }

        let task = try await AlamofireDownloadService.shared.downloadModel(model)
        return task.progress
    }

    /// Download a model with a completion handler
    static func downloadModel(
        _ modelId: String,
        progressHandler: @escaping (Double) -> Void
    ) async throws {
        let progressStream = try await downloadModel(modelId)

        for await progress in progressStream {
            progressHandler(progress.overallProgress)
            if progress.stage == .completed {
                break
            }
        }
    }
}

// MARK: - Storage Extensions

public extension RunAnywhere {

    /// Get storage information
    /// Business logic is in C++ via CppBridge.Storage
    static func getStorageInfo() async -> StorageInfo {
        return await CppBridge.Storage.shared.analyzeStorage()
    }

    /// Check if storage is available for a model download
    static func checkStorageAvailable(for modelSize: Int64, safetyMargin: Double = 0.1) -> StorageAvailability {
        return CppBridge.Storage.shared.checkStorageAvailable(modelSize: modelSize, safetyMargin: safetyMargin)
    }

    /// Get storage metrics for a specific model
    static func getModelStorageMetrics(modelId: String, framework: InferenceFramework) async -> ModelStorageMetrics? {
        return await CppBridge.Storage.shared.getModelStorageMetrics(modelId: modelId, framework: framework)
    }

    /// Clear cache
    static func clearCache() async throws {
        try SimplifiedFileManager.shared.clearCache()
        // Emit via C++ event system
        CppBridge.Events.emitStorageCacheCleared(freedBytes: 0)
    }

    /// Clean temporary files
    static func cleanTempFiles() async throws {
        try SimplifiedFileManager.shared.cleanTempFiles()
        // Emit via C++ event system
        CppBridge.Events.emitStorageTempCleaned(freedBytes: 0)
    }

    /// Delete a stored model
    /// - Parameters:
    ///   - modelId: The model identifier
    ///   - framework: The framework the model belongs to
    static func deleteStoredModel(_ modelId: String, framework: InferenceFramework) async throws {
        try SimplifiedFileManager.shared.deleteModel(modelId: modelId, framework: framework)
        // Mark the model as not downloaded (localPath: nil)
        try await CppBridge.ModelRegistry.shared.updateDownloadStatus(modelId: modelId, localPath: nil)
        // Emit via C++ event system
        CppBridge.Events.emitModelDeleted(modelId: modelId)
    }

    /// Get base directory URL
    static func getBaseDirectoryURL() -> URL {
        SimplifiedFileManager.shared.getBaseDirectoryURL()
    }

    /// Get all downloaded models
    static func getDownloadedModels() -> [InferenceFramework: [String]] {
        SimplifiedFileManager.shared.getDownloadedModels()
    }

    /// Check if a model is downloaded
    @MainActor
    static func isModelDownloaded(_ modelId: String, framework: InferenceFramework) -> Bool {
        SimplifiedFileManager.shared.isModelDownloaded(modelId: modelId, framework: framework)
    }
}
